subsys/testsuite/ztest/include/ztress.h

```diff
@@ -0,0 +1,185 @@
+
+/** @brief Declare thread execution context.
+ *
+ * Context is executed in the thread context. Priority of the thread is determined
+ * based on order in which contexts are setup in @ref ZTRESS_EXECUTE.
+ *
+ * @note Thread sleeps for random amount of time. Additionally, thread busy-waits
+ * for random time to randomize test even further.
+ *
+ * @param handler User handler of type @ref ztress_handler.
+ *
+ * @param user_data User data passed to the @p handler.
+ *
+ * @param exec_cnt Number of handler executions to complete the test. If 0 then
+ * this is not included in completion criteria.
+ *
+ * @param preempt_cnt Number of preemptions of that context to complete the test.
+ * If 0 then this is not included in completion criteria.
+ *
+ * @param init_timeout Initial backoff time base (given in @ref k_timeout_t). It
+ * is adjusted during the test to optimize CPU load. The actual timeout used for
+ * sleeping is randomized.
+ */
+#define ZTRESS_THREAD(handler, user_data, exec_cnt, preempt_cnt, init_timeout) \
+	(ZTRESS_ID_THREAD, handler, user_data, exec_cnt, preempt_cnt, init_timeout)
+
+/** @brief User handler
+ *
+ * @retval true continue test.
+ * @retval false end given context.
+ */
+typedef bool (*ztress_handler)(void *user_data, uint32_t cnt, bool last, int prio);
+
+/** @internal Context structure. */
+struct ztress_context_data {
+	/* Handler. */
+	ztress_handler handler;
+
+	/* User data */
+	void *user_data;
+
+	/* Minimum number of executions to complete the test. */
+	uint32_t exec_cnt;
+
+	/* Minimum number of preemptions to complete the test. */
+	uint32_t preempt_cnt;
+
+	/* Initial timeout. */
+	k_timeout_t t;
+};
+
+/** @internal Macro for initializing context data. */
+#define Z_ZTRESS_GET_HANDLER_DATA(data) \
+	{ \
+		.handler = GET_ARG_N(2, __DEBRACKET data), \
+		.user_data = GET_ARG_N(3, __DEBRACKET data), \
+		.exec_cnt = GET_ARG_N(4, __DEBRACKET data), \
+		.preempt_cnt = GET_ARG_N(5, __DEBRACKET data), \
+		.t = GET_ARG_N(6, __DEBRACKET data), \
+	}
+
+/** @internal Macro for checking if provided context is a timer context. */
+#define Z_ZTRESS_HAS_TIMER(data, ...) \
+	GET_ARG_N(1, __DEBRACKET data)
+
+/** @brief Setup and run stress test.
+ *
+ * Macro setup all contexts and runs them till completion. Test runs until all
+ * completion requirements are met or until test timeouts (use @ref ztress_set_timeout
+ * to configure timeout) or until test is aborted (@ref ztress_abort).
+ *
+ * On test completion report is printed (@ref ztress_report is called internally).
+ *
+ * @param ... List of contexts. Contexts are configured using @ref ZTRESS_TIMER
+ * and @ref ZTRESS_THREAD macros. @ref ZTRESS_TIMER can only be used as the first
+ * argument. Each thread context has assigned priority. Priority is assigned in
+ * descending order (first context has the highest priority). Number of supported
+ * thread contexts is configurable in Kconfig.
+ */
+#define ZTRESS_EXECUTE(...) do { \
```

Comment:
This seems fragile to me saying that a timer can only be the first argument. Can we at least have a dedicated argument for it?

Reply:
I've added build time assert which will fail if there is any `ZTRESS_TIMER` context present except for the first argument. Imo this should be enough. If we go for dedicated argument then we would need to have two macros (`ZTRESS_EXECUTE_WITH_TIMER`?) or accept that it will be empty when timer is not used (`ZTEST_EXECUTE(, ZTRESS_THREAD(...))`). In my opinion having one macro with compile time check and clear message what is wrong is better.

Reply:
Thoghts about extensibility? I agree that as it stands it'll work just fine. I'm worried about running into issues like we did with power management where the PM callback was added to the device declare directly and now any changes force us to change ~400 call sites of the device declare macro. I'm usually in the camp of splitting up the responsibilities of the macros as possible.

Just to be clear, I thought about this and couldn't come up with anything better. Though I'm not great with macros, so I was just hoping to make sure we look at this.

Reply:
It won't be as widely used as PM and we will adapt as more use cases come. I have 3 tests that i will convert to use it (one already converted: https://github.com/zephyrproject-rtos/zephyr/pull/41290) and current functionality will cover for that.